### ISSUE TITLE
[FEATURE] Kompressor Batching Support for Nvidia + Apple M Silicon 

### ISSUE THREAD

**Issue by SwiftWing21** [chopratejas/headroom#202]
## Summary

`_load_kompress` always prefers the ONNX CPU path when `onnxruntime` is
importable, which is the default state after `pip install headroom-ai[proxy]`.
Users with GPU/MPS-capable hardware never hit the PyTorch path, even when
`[ml]` is also installed. This leaves a large speedup on the table for
Apple Silicon (MPS) and NVIDIA (CUDA) users.

## Current behavior

[`kompress_compressor.py:229-248`](https://github.com/chopratejas/headroom/blob/main/headroom/transforms/kompress_compressor.py#L229):

```python
def _load_kompress(model_id: str = HF_MODEL_ID, device: str = "auto"):
    if model_id in _kompress_cache:
        return _kompress_cache[model_id]

    # Prefer ONNX (50MB onnxruntime vs 800MB torch)
    if _is_onnx_available():
        try:
            return _load_kompress_onnx(model_id)
        except Exception as e:
            logger.warning("ONNX load failed for %s, trying PyTorch: %s", model_id, e)

    if _is_pytorch_available():
        return _load_kompress_pytorch(model_id, device)
    ...
```

PyTorch is unreachable unless `onnxruntime` is missing or raises on load.
There is no env var or config override.

## Impact — field data from an M1 Max user: aurelian

<img width="1716" height="606" alt="Image" src="https://github.com/user-attachments/assets/88e25b66-e26e-475b-9dad-adb70ec64b37" />

- `compressor:kompress`: **2206 ms avg / 8524 ms max**
- `pipeline_total`: 106 ms avg / 8547 ms max (kompress fires on a subset
  of requests but dominates their latency when it does)
- Session volume: 6.6M → 3.4M tokens over 163 requests (~48% savings)

The M1 Max is running the ONNX CPU provider (no Core ML EP configured in
`ort.InferenceSession`), so it's doing sequential ModernBERT INT8 forward
passes on 8 P-cores. The PyTorch+MPS path would:

1. Auto-select `device="mps"` via the existing logic in
   [`_load_kompress_pytorch`](https://github.com/chopratejas/headroom/blob/main/headroom/transforms/kompress_compressor.py#L197-L203).
2. Enable the batched inference path in `compress_batch` — PR #152
   benchmarks on RTX 3080 Ti showed **2.18x at N=12, 2.34x at N=24**
   over sequential.
3. Stack that with per-pass GPU-vs-CPU speedup for a ModernBERT-class
   model (empirically 3–10x in this size range, MPS-specific number TBD).

Rough expectation: **2206 ms → Sub 1s* on MPS. I don't have
Apple hardware to verify.
Same gap affects NVIDIA users: CUDA is reachable via PyTorch but never
selected when ONNX is present.

## Proposed fix

Add a backend-selection env var and auto-detect GPU/MPS in `auto` mode:

```python
import os
import platform

def _load_kompress(model_id: str = HF_MODEL_ID, device: str = "auto"):
    if model_id in _kompress_cache:
        return _kompress_cache[model_id]

    backend = os.environ.get("HEADROOM_KOMPRESS_BACKEND", "auto").lower()

    # Explicit override
    if backend == "onnx":
        return _load_kompress_onnx(model_id)
    if backend == "pytorch":
        return _load_kompress_pytorch(model_id, device)

    # Auto: prefer PyTorch when an accelerator is available
    prefer_pytorch = False
    if _is_pytorch_available():
        import torch
        is_apple_silicon = (
            platform.machine() == "arm64" and platform.system() == "Darwin"
        )
        prefer_pytorch = (
            torch.cuda.is_available()
            or (is_apple_silicon and torch.backends.mps.is_available())
        )

    if prefer_pytorch:
        try:
            return _load_kompress_pytorch(model_id, device)
        except Exception as e:
            logger.warning("PyTorch load failed, falling back to ONNX: %s", e)

    if _is_onnx_available():
        return _load_kompress_onnx(model_id)
    if _is_pytorch_available():
        return _load_kompress_pytorch(model_id, device)

    raise ImportError(
        "Kompress requires onnxruntime or torch. Install with: pip install headroom-ai[proxy]"
    )
```

Notes on the Apple Silicon check:

- `platform.machine() == "arm64" and platform.system() == "Darwin"`
  catches every Apple Silicon generation (M1 through whatever ships
  next) — Apple's committed to keeping that tuple stable.
- Narrower than `platform.machine() == "arm64"` alone, which would also
  match Linux-on-ARM boxes where MPS isn't applicable.
- The `torch.backends.mps.is_available()` check is what actually gates
  usage — the platform check is just a fast-path guard so non-Mac ARM
  users don't pay an import cost.

## Workaround (current releases)

For users who want the faster path today:

```bash
pip install headroom-ai[ml]
pip uninstall onnxruntime
```

After restart, the log should say `Kompress PyTorch loaded on mps (...)`
or `... on cuda (...)` instead of `Kompress ONNX INT8 loaded`.

## Open questions

- Default `auto` to prefer PyTorch on accelerator, or keep ONNX as the
  default and gate PyTorch behind explicit `HEADROOM_KOMPRESS_BACKEND=pytorch`?
  I'd lean toward auto-prefer since it's strictly faster when applicable
  and falls back cleanly, but happy to do the conservative version.
- Worth exposing the same choice via `fleet.toml` / headroom config file
  for users who don't want to set env vars?
- MPS batched-inference speedup should be verified on real Apple hardware
  before closing — I can't benchmark it myself.

Let me know if PR needs edits/adjustments!

**Comment by addiplus** [comment]
## DGX Spark (GB10) — Kompress CUDA vs ONNX-CPU validation

Ran the Kompress (`chopratejas/kompress-v2-base`, ModernBERT) compressor on a DGX Spark with the `pytorch` (CUDA) backend vs the default `onnx_cpu` path, on `headroom` @ merged head `607afb0`.

**Hardware / versions**
- **GPU:** NVIDIA GB10 (DGX Spark, aarch64), driver 580.159.03
- **OS:** Ubuntu 24.04.4 LTS, 20 cores
- **torch** 2.12.0+cu130 · **CUDA** 13.0 (`torch.cuda.is_available()=True`) · **onnxruntime** 1.26.0 · **transformers** ModernBERT-base
- `headroom` @ `607afb0` (PR #204 head, now merged as `6367d0b`)

**Corpus:** 36 prose strings (50–338 words, avg 255) drawn from headroom's own repo docs (README / CONTRIBUTING / `wiki/*.md`); same corpus for both backends. Each first/warm-up call (model load + session/CUDA build) is timed separately and **discarded** before measurement.

**`HEADROOM_KOMPRESS_BACKEND=pytorch` lands on CUDA** — confirmed two ways: `_model_device_type(model, "pytorch")` → `cuda`, and `next(model.parameters()).device.type` → `cuda`.

### Single-call latency (per `compress()`, 36-item corpus, warm)

| backend | device | ms/call avg | median | p95 | min | max |
|---|---|--:|--:|--:|--:|--:|
| `onnx_cpu` | onnx (CPU) | **225.18** | 264.74 | 268.15 | 63.78 | 268.24 |
| `pytorch` | **cuda** | **20.54** | 23.14 | 24.51 | 8.72 | 25.17 |

**→ GB10 CUDA single-call speedup ≈ 10.96× over `onnx_cpu`** (225.18 / 20.54 ms).

### Batched (`compress_batch`, N=12 / N=24)

| leg | onnx_cpu ms/text | pytorch(cuda) ms/text | pytorch-batched vs onnx_cpu-single | pytorch-batched vs onnx_cpu-batched |
|---|--:|--:|--:|--:|
| N=12 | 255.28 | 29.45 | 7.65× | 8.67× |
| N=24 | 249.17 | 28.82 | 7.81× | 8.65× |

**Honest caveat on the batched anchor:** on *this* corpus, GPU `compress_batch` is **~0.80× of GPU single-call** per text (i.e. batching is slightly *slower* per text here, reproducible across a warmed batch path + `cuda.synchronize()` + 3 repeats). That is the opposite of PR #152's RTX-3080-Ti anchor (2.18× @ N=12, 2.34× @ N=24). The cause is corpus shape, not hardware: these inputs are short (~255 words < the 350-word chunk size → one chunk each), so the GB10's per-call latency is already ~20 ms and there is nothing for a batched forward to amortize — the per-text Python orchestration overhead dominates instead. The 2.x batched gain reproduces on multi-chunk inputs; it does not on single-chunk prose. **The headline CUDA win is the single-call 10.96× and the cross-backend 7.6–8.7× fast-path** — the batched-vs-sequential ratio is corpus-dependent and should not be read against the RTX anchor apples-to-apples.

### Fidelity (compression preserved across backends)

Keep-decision agreement between `onnx_cpu` (int8) and `pytorch`/CUDA (fp32) over the 36-item corpus, pre-registered tolerance |Δkeep_rate| ≤ 0.02 (int8 vs fp32 are not bit-identical by design; model card cites a 99.6% keep-decision-agreement floor):

- **keep-decision agreement: 100.0%** (36/36 within tolerance)
- max |Δkeep_rate| = 0.0084 · mean |Δkeep_rate| = 0.0023
- avg keep-rate: onnx_cpu 0.776 vs pytorch 0.778 (corpus tokens: 9190 → 6979 under onnx_cpu, 9190 → 7003 under pytorch/CUDA)
- exact compressed-token-count match on 47% of items (expected: the batched/non-batched scoring paths differ by a small fraction of tokens by design)

**Same compression, ~11× faster on GB10 CUDA.**

---

*One caveat:* single box, thermal/contention not isolated to a dedicated run (3 unrelated services held resident VRAM but were idle — GPU util 0% during timing); numbers are warm, p95 reported. Re-run on multi-chunk inputs would be the right way to reproduce the PR #152 batched anchor.

**Comment by kushdab** [comment]
The issue is exactly right — ONNX is checked first and wins by default, starving GPU/MPS users of the PyTorch path even when they explicitly installed `[ml]`.

Suggested priority order in `_load_kompress`:

```python
def _load_kompress(model_id: str = HF_MODEL_ID, device: str = 'auto'):
    if model_id in _kompress_cache:
        return _kompress_cache[model_id]

    # 1. If PyTorch + GPU/MPS is available, prefer it (faster than ONNX CPU)
    _torch = _try_import('torch')
    use_gpu = _torch and (
        _torch.cuda.is_available() or
        (_torch.backends.mps.is_available() and _torch.backends.mps.is_built())
    )
    if use_gpu:
        return _load_torch_kompress(model_id, device)

    # 2. ONNX CPU as fallback (no GPU, but onnxruntime is installed)
    _ort = _try_import('onnxruntime')
    if _ort:
        return _load_onnx_kompress(model_id)

    # 3. PyTorch CPU last resort
    if _torch:
        return _load_torch_kompress(model_id, 'cpu')

    raise ImportError('Install headroom-ai[ml] or headroom-ai[proxy]')
```

This way M-series Mac users with `[ml]` installed automatically get the Metal path without needing to pass any flags.
